recirq/otoc/otoc_circuits.py

```diff
@@ -0,0 +1,571 @@
+        light_cone_u_dagger = _extract_light_cone(
+            butterfly_qubits, cycle, interaction_sequence,
+            reverse=True, from_right=False, multiple_butterflies=multi_b)
+        light_cone_meas = _extract_light_cone(
+            qubits[0], cycle, interaction_sequence,
+            reverse=True, from_right=True, multiple_butterflies=False)
+
+        full_cone = [*light_cone_u, *light_cone_u_dagger]
+        idle_echo_indices = np.zeros((num_qubits, cycle * 2), dtype=int)
+
+        if idle_echo_seed is not None:
+            np.random.seed(idle_echo_seed)
+
+        for idx, qubit in enumerate(qubits):
+            idle_cycles = []
+            for i, cone in enumerate(full_cone):
+                if qubit not in cone:
+                    idle_cycles.append(i)
+            echo_locs = list(range(idle_echo_frequency - 1, len(idle_cycles),
+                                   idle_echo_frequency))
+            if len(echo_locs) % 2 == 1:
+                echo_locs.remove(echo_locs[-1])
+            if idle_echo_seed is None:
+                rand_echoes = np.ones(int(len(echo_locs) / 2), dtype=int) + 1
+            else:
+                rand_echoes = np.random.choice(4, int(len(echo_locs) / 2)) + 1
+            for k, echo_idx in enumerate(rand_echoes):
+                idle_echo_indices[idx, idle_cycles[echo_locs[2 * k]]] = echo_idx
+                idle_echo_indices[idx, idle_cycles[
+                    echo_locs[2 * k + 1]]] = (echo_idx + 1) % 4 + 1
+
+        echo_indices_for = idle_echo_indices[:, 0:cycle]
+        echo_indices_back = idle_echo_indices[:, cycle:(2 * cycle)]
+
+        seq_for, seq_back = _compile_interactions(
+            qubits[0], butterfly_qubits, cycle, interaction_sequence)
+        moments_for = _compile_moments(seq_for, forward_ops,
+                                       layer_by_layer_cal=layer_by_layer_cal)
+        moments_back = _compile_moments(seq_back, reverse_ops,
+                                        layer_by_layer_cal=layer_by_layer_cal)
+
+        circuits_forward, _ = build_xeb_circuits(
+            qubits, [cycle], moments_for, sq_rand_nums=rand_nums,
+            z_only=z_only, ancilla=ancilla, cycles_per_echo=cycles_per_echo,
+            light_cones=[light_cone_u], echo_indices=echo_indices_for)
+        circuits_backward, _ = build_xeb_circuits(
+            qubits, [cycle], moments_back, sq_rand_nums=rand_nums_rev,
+            z_only=z_only, ancilla=ancilla, cycles_per_echo=cycles_per_echo,
+            reverse=True, light_cones=[light_cone_u_dagger, light_cone_meas],
+            echo_indices=echo_indices_back)
+
+    butterfly_ops = [None, cirq.X, cirq.Y, cirq.Z]
+    prep_phases = [0.0, 1.0]
+    meas_phases = [0.0, 1.0]
+    otoc_circuits = []
+
+    # Combine $U$ and $U^\dagger$ with other gates related to SPAM to
+    # complete the OTOC circuits.
+    for b_ops, p_prep, p_meas in itertools.product(
+            butterfly_ops, prep_phases, meas_phases):
```

Comment:
Weird wrapping. We can use 100 character line-length

Reply:
Fixed.